design.md

```diff
@@ -0,0 +1,242 @@
+1. Resource controls of CPU, memory and disk IO using cgroups.
+
+## API
+
+1. gRPC API to start/stop/get status/stream output of a running process.
+1. Use mTLS for authentication.
+1. Has a simple authentication scheme.
+
+## Client
+
+1. CLI connects to worker service to start, stop, get status, and stream output of a job.
+2. Multiple clients can connect to the service.
+
+
+# Design
+
+## Worker Library and Server
+
+The server will use a library to: 
+
+1. Start a job. Starting a job returns a UUID which is the job id. This ID can be used to do further
+2. Stop a job. Sends SIGKILL to the process. SIGKILL instantly terminates the process.
+3. Query job.  Returns the status of a job. 
+4. Provide streaming of job logs.
+
+Service should be able to handle multiple clients. 
+
+### Tradeoffs and Production 
+- Initially using SIGTERM, then SIGKILL as a last resort is probably the better way to do things. But for this 
+  challenge, SIGKILL is sufficient.
+
+
+### cgroups
+
+`cgroups v2` will be used for resource control. 
+
+- Each job will run in its own cgroup.
+- Resource usage will be isolated per job.
+- There will be job_id grouping. for example: 
+  - /sys/fs/cgroup/teleport/<job_id> -- the job_id is the UUID provided by the server when the job was created.
+
+The hardcoded resource limits are below. For this challenge, I will assume the jobs are relatively light: 
+- CPU: 0.5 
+- Memory: 256MB
+- IO: 1MB/s
+
+#### Placing jobs in cgroups
+
+- os.Open opens /sys/fs/cgroup/teleport/<job_id>, resulting in a file descriptor.
+- When running exec.Cmd, use UseCgroupFD and CgroupFD to specify that the process should be placed in a cgroup.
+
+
+#### Tradesoffs and the Production environment
+- It is preferable to dynamically configure resource limits based on job and system state. A first step in doing 
+  this is providing a way to adjust resource limits when jobs are created.
+- In production, we probably want orchestrators, like systemd, which manages cgroups. 
+
+## API
+
+### Using gRPC for communication. 
+
+- Start: Starts a job with hardcoded resource limits.
+- Stop: Stops a job.
+- GetStatus: Fetches status of a job
+- GetLogs: Stream logs of a running job.
+
+```
+service TeleportService {
+    rpc Start (StartRequest) returns (StartResponse);
+    rpc Stop (StopRequest) returns (StopResponse);
+    rpc GetStatus (GetStatusRequest) returns (GetStatusResponse);
+    rpc GetLogs (GetLogsRequest) returns (stream GetLogsResponse);
+}
+
+message StartRequest {
+    string command = 1;
+    repeated string args = 2;
+}
+
+message StartResponse {
+    string job_id = 1; // a job UUID
+}
+
+message StopRequest {
+    string job_id = 1;
+}
+
+message StopResponse {
+    // ideally, add a boolean to see if stopping was successful
+    // maybe even a message. But not necessary for the challenge.
+}
+
+message GetStatusRequest {
+    string job_id = 1;
+}
+
+message GetStatusResponse {
+    Status status = 1;
+}
+
+enum Status{
```

Comment:
The Protobuf style guide states you should include a zero value and that this zero value should:

> The zero value enum should have the suffix UNSPECIFIED, because a server or application that gets an unexpected enum value will mark the field as unset in the proto instance. The field accessor will then return the default value, which for enum fields is the first enum value. For more information on the unspecified enum value, see [the Proto Best Practices page](https://protobuf.dev/programming-guides/dos-donts#unspecified-enum).

Reply:
addressed: https://github.com/rexposadas/teleport/pull/1/commits/dc45179d5bfb3f7ccef45a9c870aaf9d6df828dc